[util] make topgen run multiple passes

tjaychen opened the pull request:
- Temporary solution to #2083
- This is an UBER hack
- This is needed because the generation process has dependencies on itself.  Normally this can be worked around, but for tops such as top_englishbreakfast that are generated on the fly, it becomes probelmatic.
- The basic idea is that for each dependency the generation process is invoked again to make sure updated information is used to generate new correct modules.
- The final run is the one that is actually used for final merging and deployed by the rest of the script.

Signed-off-by: Timothy Chen <timothytim@google.com>

[util] slightly tweak topgen multi-pass

Signed-off-by: Timothy Chen <timothytim@google.com>

eunchan commented on util/topgen.py:
```diff
@@ -287,7 +287,6 @@ def generate_plic(top, out_path):
     with rtl_gen_path.open(mode='w', encoding='UTF-8') as fout:
         fout.write(genhdr + gencmd + out)
 
-
```
I think flake8 requires two lines between the functions.

tjaychen replied:
huh... you're right. Why didn't this fail lint..?

eunchan commented on util/topgen.py:
```diff
@@ -993,7 +879,7 @@ def main():
     for ip in generated_list:
         log.info("Appending {}".format(ip))
-        if ip == 'clkmgr':
+        if ip == 'clkmgr' or pass_idx:
```
Let's explicitly add checker `pass_idx == 1` ? or I think it is more like `pass_idx != 0`

tjaychen replied:
yeah it's meant to be > 0.

eunchan commented on util/topgen.py:
```diff
@@ -1099,6 +981,145 @@ def main():
     # These modules are not templated, but are not in hw/ip
     generate_top_only(top_only_list, out_path, topname)
 
+    if pass_idx:
```
ditto here. `pass_idx == 1` in this case. or we can pack two if s `pass_idx == 1 and args.top_ral`

eunchan commented on util/topgen.py:
I am a little bit confused here. the code updates `cfg_copy` and discard?  I thought the second pass to use the result of amended cfgs. Could you please comment?

tjaychen replied:
yeah... so .. our amend updates `topcfg`, and if this is fed back again, it actually creates more warnings / errors (didn't look into it) because the original structure has changed.

So instead what I'm doing is i keep discarding until the last run, where i actually take the results.  The previous runs just generate the correct modules.

So if you take my example..

during the first run we gather all the modules, and generate the correct pinmux.  However, the gathering used stale pinmux, so it goes back again.  

in teh second example, the first run we gather all the modules, and generate the correct pinmux.  However the pwrmgr generation is wrong.  So we go back and second run and gather correct modules including pinmux, and now we generate correct pwrmgr.  Now we go back a 3rd time and use all the correct modules for the final run.

eunchan replied:
AH.. OK Now I got the point! Thanks for explanation. I think it is worth to add your comment to the comment. (hum weird sentence but... you got the point right? :) )

LGTM

tjaychen replied:
haha sounds good. It's like a recursive comment!

vogelpi reviewed (approved):
Thanks @tjaychen for this PR. As you say it's kind of hacky but at the moment, I don't see a quick an clean solution to the problem. Thanks for commenting the code, that's useful. Nit, you may even want to add a reference in the code to GitHub issue #2083 which seems to be very related.

There is currently an older comment around multiple passes required for the PLIC. The comment was there before but I think it would be better to remove it to reduce confusion now as  we are now anyway doing multiple passes not just because of the PLIC. It's on line 874:
```
    # It may require two passes to check if the module is needed.
    # TODO: first run of topgen will fail due to the absent of rv_plic.
    # It needs to run up to amend_interrupt in merge_top function
    # then creates rv_plic.hjson then run xbar generation.
```
Related to that, it seems we have an old issue #28.

Other than that, this looks good to me!

tjaychen commented:
> Thanks @tjaychen for this PR. As you say it's kind of hacky but at the moment, I don't see a quick an clean solution to the problem. Thanks for commenting the code, that's useful. Nit, you may even want to add a reference in the code to GitHub issue #2083 which seems to be very related.
> 
> There is currently an older comment around multiple passes required for the PLIC. The comment was there before but I think it would be better to remove it to reduce confusion now as we are now anyway doing multiple passes not just because of the PLIC. It's on line 874:
> 
> ```
>     # It may require two passes to check if the module is needed.
>     # TODO: first run of topgen will fail due to the absent of rv_plic.
>     # It needs to run up to amend_interrupt in merge_top function
>     # then creates rv_plic.hjson then run xbar generation.
> ```
> 
> Related to that, it seems we have an old issue #28.
> 
> Other than that, this looks good to me!

thanks @vogelpi @eunchan 
i've updated the comments significantly and also added the reference to the issue you pointed out. 
I won't close that issue out because..i feel like this is only a temporary solution.
